### DISCUSSION TITLE
How to dynamically resize eframe window?

### DISCUSSION THREAD

**Discussion by SCJangra** [emilk/egui#4329] [Q&A]
https://docs.rs/eframe/0.18.0/eframe/struct.Frame.html#method.set_window_size is removed in the newer versions, is there any other way to resize the main eframe window?

**Comment by YgorSouza** [comment]
Use the method [Context::send_viewport_cmd](https://docs.rs/egui/0.27.2/egui/struct.Context.html#method.send_viewport_cmd) with the [ViewportCommand::InnerSize](https://docs.rs/egui/0.27.2/egui/viewport/enum.ViewportCommand.html#variant.InnerSize) variant.

**Reply by SCJangra** [reply]
@YgorSouza thank you for the quick response, I am trying to restrict the eframe window size to the size of an egui Window. I've tried the following

```rust
use eframe::egui;

fn main() {
    let native_options = eframe::NativeOptions {
        viewport: egui::ViewportBuilder::default().with_resizable(false),
        ..Default::default()
    };
    eframe::run_native(
        "My egui App",
        native_options,
        Box::new(|cc| Box::new(MyEguiApp::new(cc))),
    )
    .unwrap();
}

#[derive(Default)]
struct MyEguiApp {}

impl MyEguiApp {
    fn new(_cc: &eframe::CreationContext<'_>) -> Self {
        Self::default()
    }
}

impl eframe::App for MyEguiApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::Window::new("Some window")
            .anchor(egui::Align2::LEFT_TOP, [0.0, 0.0])
            .show(ctx, |ui| {
                ui.heading("Hello world!");
            });

        let size = ctx.used_size();
        ctx.send_viewport_cmd(egui::ViewportCommand::InnerSize(size));
    }
}
```

![image](https://github.com/emilk/egui/assets/40860699/fd9c918c-9e41-4cb4-ab65-e4bdea66f7d5)

This does resize the eframe window, but for some reason it shrinks the egui window to the bottom left corner. Maybe the `update` function is not the right place to resize the eframe window. Would you be able to tell me how this can be done?

The window does not have any borders or title bar because I've disabled those in KDE.

Thanks

**Reply by YgorSouza** [reply]
I'm not sure what is happening on your screenshot, but when I run your code the window starts flickering between different sizes. There are a few things:

- If you want to resize the window, you shouldn't put `with_resizable(false)`.
- You can disable the system decorations from eframe, with `with_decorations(false)` (although I haven't tested that on KDE since I use Gnome).
- You have to put something to fill the empty space in the egui Window, otherwise you won't be able to resize it. Like in [this example](https://github.com/emilk/egui/blob/78d95f430b09e3040d3f383fde8fea2ae1592afb/crates/egui_demo_lib/src/demo/tests.rs#L581-L594) from the demo.

This code works a little better:

<details><summary>Details</summary>
<p>

```rs
use eframe::egui;

fn main() {
    let native_options = eframe::NativeOptions {
        viewport: egui::ViewportBuilder::default()
            .with_decorations(false)
            .with_transparent(true),
        ..Default::default()
    };
    eframe::run_native(
        "My egui App",
        native_options,
        Box::new(|cc| Box::new(MyEguiApp::new(cc))),
    )
    .unwrap();
}

#[derive(Default)]
struct MyEguiApp {}

impl MyEguiApp {
    fn new(_cc: &eframe::CreationContext<'_>) -> Self {
        Self::default()
    }
}

impl eframe::App for MyEguiApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        let mut open = true;
        egui::Window::new("Some window")
            .scroll(true)
            .open(&mut open)
            .anchor(egui::Align2::LEFT_TOP, [0.0, 0.0])
            .show(ctx, |ui| {
                ui.heading("Hello world!");
                ui.allocate_space(ui.available_size());
            });

        let size = ctx.used_size();
        ctx.send_viewport_cmd(egui::ViewportCommand::InnerSize(size));
        if !open {
            ctx.send_viewport_cmd(egui::ViewportCommand::Close);
        }
    }
}

```

</p>
</details> 

But it's still not great, because there is some lag when increasing the window size. The egui window is clamped to the native window size, so clamping the native window to the egui window creates a feedback loop that means it can only increase by a couple of pixels per frame when you drag it. Also it cannot be moved. I'm not sure what would be the best way to implement that. You can take a look at the [custom_window_frame](https://github.com/emilk/egui/tree/78d95f430b09e3040d3f383fde8fea2ae1592afb/examples/custom_window_frame) example, which is close to what you are trying to do. It handles moving, but not resizing. Maybe you can combine it with your code and come up with something that works for you.

**Reply by SCJangra** [reply]
> If you want to resize the window, you shouldn't put with_resizable(false).

I am trying to create a fixed-size dialog box, with a single line text input, so I don't want it to be resizable.

> You can disable the system decorations from eframe, with with_decorations(false) (although I haven't tested that on KDE since I use Gnome).

I wasn't actually trying to create a decoration-less window. I just don't like decorations (except shadows) in general, so I try to disable those in all desktop environments/window managers I use. They are always disabled globally.

I managed to do what I wanted using this code. I know it's not pretty, but it works until I find something better.

```rust
use eframe::egui;

fn main() {
    let mut app = App::default();

    // Compute window size.
    let ctx = egui::Context::default();
    let _ = ctx.run(egui::RawInput::default(), |_| {});
    app.render(&ctx);
    let size = ctx.used_size();

    let native_options = eframe::NativeOptions {
        viewport: egui::ViewportBuilder::default()
            .with_resizable(false)
            .with_inner_size(size),
        ..Default::default()
    };
    eframe::run_native("App", native_options, Box::new(|_| Box::new(app))).unwrap();
}

#[derive(Default)]
struct App {
    input: String,
}

impl App {
    fn render(&mut self, ctx: &egui::Context) {
        egui::Area::new(egui::Id::new("new-area"))
            .fixed_pos((0.0, 0.0))
            .show(ctx, |ui| {
                ui.label("Enter text");
                ui.text_edit_singleline(&mut self.input);
            });
    }
}

impl eframe::App for App {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        self.render(ctx)
    }
}
```

Thanks for the help.
